src/chain.h

```diff
@@ -162,14 +162,16 @@ class CBlockIndex
     //! height of the entry in the chain. The genesis block has height 0
     int nHeight{0};
 
+    const std::shared_ptr<Mutex> cs_data = std::make_shared<Mutex>();
```

Comment:
Huh? Does this create a mutex for every block? What is the memory footprint?

Reply:
yeah, good topic. I was so focused on accessing disk concurrently that forgot that we have 800k block indexes loaded in memory all the time.
The mutex size is 64 bytes, so.. a mutex per block index is a no go.

Could move it to a single global mutex for block indexes file data access, which would still decouple it from `cs_main` (which is a good step forward) but.. it's still not the best solution for concurrent access.

Reply:
Working on it 👌🏼

Reply:
^ is this done, or still in progress?

Reply:
> ^ is this done, or still in progress?

Yeah, have pushed the readers/writer lock in the last update.

Still, have to say that code elegancy wise, I'm not totally happy with the `sync.h` shared lock wrapper. Planning to give it another look later this week.

Reply:
Why not do a mutex per blockfile (rather than CBlockIndex entry, which would indeed be crazy)? There are hundreds to thousands of those, so keeping a map of blockfile number to mutex shouldn't be unreasonable.

Reply:
While that would be better than the initial per `CBlockIndex` entry mutex, it's not as good as a single RW lock (which is the current PR state).

Essentially, we aim to allow concurrent access for read-only operations, while write operations that are not performed regularly require exclusive access (which is the RW lock definition).

To place it more in our code. The only time where the `CBlockIndex` file position can be modified during the entire software lifecycle is pruning, and it's done only once. And because of that single specific scenario that does not happen often, we have been forced to lock `cs_main` every time that any process wants to access the block index file position.

So, guarding this single specific case with an exclusive lock, the rest of the read-only operations can be processed concurrently by sharing ownership of the readers-only lock.

Reply:
A little add to my last comment:

I think that if we ever add other use cases that modify the block index file position more frequently (like re-ordering/compacting block data on disk or re-downloading blocks), then a more fine-grained RW lock per block file would be a nice idea.

-- the comment came to my mind because one of my long-term projects introduces the pruned blocks re-download feature.. --